### PR TITLE
cmake: some fixes and suppressions for WITH_ASAN under vstart.sh

### PR DESCRIPTION
**Pull request by cbodley**
* `boost::context` understands a `BOOST_USE_ASAN` define to add asan annotations to its context switches. BuildBoost.cmake adds this define when our cmake variable `WITH_ASAN` is set

* src/ceph.in automatically adds libasan.so to `LD_PRELOAD`, but its detection logic in `FindSanitizers.cmake` wasn't looking for newer versions of the library. on fedora 37, i have libasan.so.8 installed

* memory leaks in `bin/ceph-mon` and the `bin/ceph` prevented vstart.sh from bootstrapping the cluster, so a `qa/lsan.supp` file was added to suppress them. the suppressions are enabled with the environment variable `LSAN_OPTIONS="suppressions=../qa/lsan.supp"`

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cbodley** [comment]
tested only with gcc using `WITH_ASAN=ON WITH_UBSAN=ON`

File: qa/lsan.supp [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+# from src/ceph.in
+leak:^PyObject_Malloc
+leak:^_PyObject_Realloc
+leak:^PyMem_Calloc
+leak:^PyMem_Malloc
+leak:^PyUnicode_New
+leak:^_PyBytes_FromSize
+leak:^_PyObject_GC_Alloc
+leak:^_PyObject_GC_Resize
```
Comment (yuvalif) [inline qa/lsan.supp:17]: had to add this (using libc and rhel8):
```
leak:__interceptor_malloc
```

Comment (cbodley) [reply qa/lsan.supp:17]: `__interceptor_malloc` is the asan hook that *catches* the leaks. i think you'll want to suppress the function that calls that, instead. ex:
```
Indirect leak of 22079 byte(s) in 23 object(s) allocated from:                                                                                                
    #0 0x7f82e4eba6af in __interceptor_malloc (/usr/lib64/libasan.so.8+0xba6af)                                                                               
    #1 0x7f82e4b03419 in PyObject_Malloc (/lib64/libpython3.8.so.1.0+0x103419)
```
do you have the LeakSanitizer output for the leak you saw? maybe it's the same rgw_admin.cc leak from https://github.com/ceph/ceph/pull/50383?

**Comment by yuvalif** [comment]
had to add the following option:
```
ASAN_OPTIONS="detect_odr_violation=0"
```
to avoid:
```
==2841973==ERROR: AddressSanitizer: odr-violation (0x5624b9734a40):
  [1] size=16 'alloc_pgmap_inc' ../src/mon/PGMap.cc:42:1
  [2] size=16 'alloc_pgmap_inc' ../src/mon/PGMap.cc:42:1
```

**Comment by cbodley** [comment]
with the boost::context commit, i was hoping to get rid of these `rgw::notify::Manager` errors: https://gist.github.com/cbodley/a5d7b8f74f353bae6156c32076a8f281

but with boost::context compiled with asan and ucontext, i get a different one: https://gist.github.com/cbodley/d0ab0a9a75fcfeb5341e3a09f80eee5a

**Comment by yuvalif** [comment]
> with the boost::context commit, i was hoping to get rid of these `rgw::notify::Manager` errors: https://gist.github.com/cbodley/a5d7b8f74f353bae6156c32076a8f281
> 
> but with boost::context compiled with asan and ucontext, i get a different one: https://gist.github.com/cbodley/d0ab0a9a75fcfeb5341e3a09f80eee5a

i see the `rgw::notify::Manager` errors only in radosgw-admin.
maybe some shutdown issue?

File: qa/lsan.supp (hunk 1 above)
Comment (idryomov) [reply qa/lsan.supp:17]: > do you have the LeakSanitizer output for the leak you saw? maybe it's the same rgw_admin.cc leak from #50383?

Possibly related concern: I don't see much overlap with https://github.com/python/cpython/blob/3.11/Misc/valgrind-python.supp.  Are we sure it's not our stuff in `src/pybind` that needs fixing?

Comment (cbodley) [reply qa/lsan.supp:17]: > Are we sure it's not our stuff in `src/pybind` that needs fixing?

good question, i wasn't able to tell one way or the other from that leak output. i might try running some of the tests under `src/test/pybind/` to see if they show up there too

in general we should only suppress issues from ceph's dependencies, and open tracker issues to get ceph bugs fixed

Comment (cbodley) [reply qa/lsan.supp:17]: some interesting results.. i tried running `python -m radossss` against several different versions of python. the command immediately fails with `No module named radossss`, and doesn't import any of ceph's modules

https://gist.github.com/cbodley/07a84ad5acf9bb0c8e956f7a6204e659

each python version after 3.8 shows fewer leaks, until 3.12 shows none. so i don't think any of the python leaks i put into `lsan.supp` are specific to stuff under `src/pybind`

**Comment by yuvalif** [comment]
> with the boost::context commit, i was hoping to get rid of these `rgw::notify::Manager` errors: https://gist.github.com/cbodley/a5d7b8f74f353bae6156c32076a8f281
> 
> but with boost::context compiled with asan and ucontext, i get a different one: https://gist.github.com/cbodley/d0ab0a9a75fcfeb5341e3a09f80eee5a
> 
> edit: also tried with `ASAN_OPTIONS="detect_stack_use_after_return=true"` and it got further, but crashed instead: https://gist.github.com/cbodley/9a98541c82664efd13ca3cfc19556d5b

this PR: https://github.com/ceph/ceph/pull/50403 should fix these errors

**Review by idryomov — APPROVED** [review]
Not a fan of `NAMES` list in `cmake/modules/FindSanitizers.cmake` but we already touched on that in https://github.com/ceph/ceph/pull/50320#discussion_r1125722246.  Otherwise LGTM!

**Comment by cbodley** [comment]
> Not a fan of `NAMES` list in `cmake/modules/FindSanitizers.cmake`

i could imagine vstart.sh and vstart_environment.sh exporting the necessary LD_PRELOAD/ASAN_OPTIONS/LSAN_OPTIONS

that could avoid the `ASAN_LIBRARY` logic for `src/ceph.in`, but would only cover vstart clusters - other deployments would have to add these options manually

using `ldd` to discover the libasan so version (if it's enabled at all) would be easier to do in vstart, where we know the location of our ceph binaries. in the `ceph` python command, it's less clear what we'd run `ldd` on

personally, i'm most interested in enabling asan for a) local vstart testing, and b) jenkins 'make check'. what do you guys think?

**Comment by yuvalif** [comment]
I'm trying on fedora 37. with `libasan-12.2.1-4.fc37.x86_64` (and `/usr/lib64/libasan.so.8`)
after running:
```
cmake -DWITH_ASAN=ON -DWITH_UBSAN=ON ..
```
i get a cmake error:
```
-- Performing Test Sanitizers_ARE_SUPPORTED
-- Performing Test Sanitizers_ARE_SUPPORTED - Failed
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Sanitizers (missing: Sanitizers_ARE_SUPPORTED)
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindSanitizers.cmake:63 (find_package_handle_standard_args)
  CMakeLists.txt:603 (find_package)
```

File: qa/lsan.supp [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+# from src/ceph.in
+# leaks are specific to older python versions, and go away in 3.12
+leak:^PyObject_Malloc
+leak:^_PyObject_Realloc
+leak:^PyMem_Calloc
+leak:^PyMem_Malloc
+leak:^PyUnicode_New
+leak:^_PyBytes_FromSize
+leak:^_PyObject_GC_Alloc
+leak:^_PyObject_GC_Resize
```
Comment (yuvalif) [inline qa/lsan.supp:18]: this time the correct functions that should be added when running with rhel8 (Python 3.6.8):
```
leak:^_PyObject_Alloc
leak:^PyType_GenericAlloc
leak:^PyBytes_FromStringAndSize
leak:^set_table_resize
```

**Comment by cbodley** [comment]
> -- Performing Test Sanitizers_ARE_SUPPORTED
> -- Performing Test Sanitizers_ARE_SUPPORTED - Failed

this would be coming from `check_cxx_source_compiles("int main() {}" Sanitizers_ARE_SUPPORTED)`

you should be able to find the compilation error message from that in build/CMakeFiles/CMakeError.log

**Comment by idryomov** [comment]
> in the `ceph` python command, it's less clear what we'd run `ldd` on

I suggested running it on `rados` command in the same directory.  `rados` should always be present if `ceph` is, right?

**Comment by cbodley** [comment]
> I suggested running it on `rados` command in the same directory. `rados` should always be present if `ceph` is, right?

oops, i had assumed that `rados` was also a python script 👍 so the vstart piece would only need to add any ASAN_OPTIONS/LSAN_OPTIONS

**Comment by yuvalif** [comment]
> build/CMakeFiles/CMakeError.log

no error file

**Comment by yuvalif** [comment]
ASAN found another issue. fixed here: https://github.com/ceph/ceph/pull/50425

**Comment by cbodley** [comment]
> ASAN found another issue. fixed here: #50425

that's a dup of https://github.com/ceph/ceph/pull/50383, sorry. i had linked that in a comment above

**Comment by yuvalif** [comment]
> > with the boost::context commit, i was hoping to get rid of these `rgw::notify::Manager` errors: https://gist.github.com/cbodley/a5d7b8f74f353bae6156c32076a8f281
> > but with boost::context compiled with asan and ucontext, i get a different one: https://gist.github.com/cbodley/d0ab0a9a75fcfeb5341e3a09f80eee5a
> > edit: also tried with `ASAN_OPTIONS="detect_stack_use_after_return=true"` and it got further, but crashed instead: https://gist.github.com/cbodley/9a98541c82664efd13ca3cfc19556d5b
> 
> this PR: #50403 should fix these errors

only fixed radosgw-admin issue. not the RGW issue :-(

**Comment by cbodley** [comment]
passed rgw qa in https://pulpito.ceph.com/cbodley-2023-03-10_19:16:19-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**